### ISSUE TITLE
[BUG] - Initial sync may block forever with CPU maxed out

### ISSUE THREAD

**Issue by tdiesler** [input-output-hk/cardano-node#2251]
**External**

**Initial Sync**

**Summary**

After a varying number of slots the node freezes with CPU maxed out. 

* There is no more logging to console nor to the (debug) log file
* The EKG/Prometheus endpoints do not respond any more
 
As if the node has gotten itself in a tight endless loop.

**Steps to reproduce**

On ARM64 I do ...

```
docker run --detach \
    --name=relay \
    -p 3001:3001 \
    nessusio/cardano run

docker logs -f relay
```
https://hub.docker.com/r/nessusio/cardano

**Expected behavior**

Initially, I'd like to see some indication of what is going on in the log so that eventually we can get to the bottom of this.

**System info**
- ARM64
- RaspberryPi4 or AWS CentOS8 
- Docker
- cardano-node-1.24.2

**Screenshots and attachments**

Nothing suspicious in the log

```
[relay:cardano.node.ChainDB:Info:43] [2021-01-04 19:21:41.82 UTC] Valid candidate 3f8a7c65f73ee0485dcad85bc2174c0eb7ab651bcf5110721b6fb8613ca0a73e at slot 44576
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:41.82 UTC] Chain extended, new tip: 3f8a7c65f73ee0485dcad85bc2174c0eb7ab651bcf5110721b6fb8613ca0a73e at slot 44576
[relay:cardano.node.ChainDB:Info:43] [2021-01-04 19:21:41.83 UTC] Valid candidate 59d9848454e1b12dcb985695fe424d9823ebc0f0ac5d919b579ea8e972c663fa at slot 44577
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:41.84 UTC] Chain extended, new tip: 59d9848454e1b12dcb985695fe424d9823ebc0f0ac5d919b579ea8e972c663fa at slot 44577
[relay:cardano.node.ChainDB:Info:579] [2021-01-04 19:21:41.87 UTC] before next, messages elided = 38760154499862
[relay:cardano.node.ChainDB:Info:579] [2021-01-04 19:21:41.87 UTC] Valid candidate 92f47f9867cc1a2e0e6e2ec4f04a1fcb7764cc395cb0a00447823b907c59de3f at slot 44578
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:41.87 UTC] Chain extended, new tip: 92f47f9867cc1a2e0e6e2ec4f04a1fcb7764cc395cb0a00447823b907c59de3f at slot 44578
[relay:cardano.node.ChainDB:Info:43] [2021-01-04 19:21:41.97 UTC] Valid candidate 7ea7a82ce1d7ae747532b7174a4fc03f022a99e905bdffdc4aa95bb9f13b2903 at slot 44588
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:41.97 UTC] Chain extended, new tip: 7ea7a82ce1d7ae747532b7174a4fc03f022a99e905bdffdc4aa95bb9f13b2903 at slot 44588
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:42.01 UTC] Chain extended, new tip: 6e2dedc9f5188ce372908ea5b31bccc69fba5d231928ce9bd924580bf7f97a47 at slot 44593
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:42.10 UTC] Chain extended, new tip: 506678c536215bebde58dbec021eae2c6f1485c6273d56a25829da4335909aae at slot 44605
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:42.13 UTC] Chain extended, new tip: df0824814a70181c39230bc5e5c896d1c84a83659743eccd474f0a045f493f87 at slot 44609
[relay:cardano.node.ChainDB:Info:43] [2021-01-04 19:21:42.17 UTC] Valid candidate f752a1fe0095404a9170ec32a113fb3a87e04d6f5090de3c7fd102c827fe0212 at slot 44612
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:42.17 UTC] Chain extended, new tip: f752a1fe0095404a9170ec32a113fb3a87e04d6f5090de3c7fd102c827fe0212 at slot 44612
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:42.21 UTC] Chain extended, new tip: 9b68f553ff387583d28b9393e7038628d012fa4a747c58a5d4a709189ca5da22 at slot 44614
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:42.23 UTC] Chain extended, new tip: 0776a4927813c3bb2ad110399d7e42df5e7427041ab221aab41d1a5007bff094 at slot 44617
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:43.48 UTC] Chain extended, new tip: 1e90177828e3433851ea7a62c0e40c1d08df5efebe3127735c22928f8c1b6a8f at slot 44820
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:44.73 UTC] Chain extended, new tip: b4a0943d858cb6b60a4e9ebc23b8e7e77bfdc4ac74094fcdced7a3b436eaaec5 at slot 45022
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:45.99 UTC] Chain extended, new tip: 5e3981dec27f84bd98e2ea155916ce4730e1a9f83361d57acfcb84c5206b4a01 at slot 45242
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:47.24 UTC] Chain extended, new tip: 475f393f3797e771cd5639cdc8cef96f1f15ae3bbe7ca07055d00540c07c89e6 at slot 45463
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:48.49 UTC] Chain extended, new tip: 3137b378f9c43c658473b03603c1e0c7893969f32220cc5d8e62e33a124ecae0 at slot 45628
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:49.75 UTC] Chain extended, new tip: f607d1f38508a965da678c941bd3c4d4aae57fac6e682872d4ae1c1aca788cbe at slot 45765
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:51.00 UTC] Chain extended, new tip: 165c54de8e7b37c8fa7455e60d5fd566d38305e11e85400999043192fad3723b at slot 45930
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:52.25 UTC] Chain extended, new tip: 3cab325f853b82b54fb9564c1f5126b932f08ffac96960605f650780bb0a3a79 at slot 46123
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:53.50 UTC] Chain extended, new tip: ac15ef556795e551043b0851f96bf4daa69241910cd3e44e86c2ade4f7c3c2f2 at slot 46325
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:54.75 UTC] Chain extended, new tip: feca1e1be67f1f3a83fa644f7bd4e4a660facae96f6014bff90e7decbd03de61 at slot 46554
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:56.00 UTC] Chain extended, new tip: 846c6ce3c92b27c91b2385dfc15719c1f691e1ff20dd6049a9fc3feeaa48c093 at slot 46779
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:57.26 UTC] Chain extended, new tip: 681ced6441ee47ab7e53ce0ca2d19794ad20ae7a55d563b4acdf94604811d837 at slot 46925
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:58.52 UTC] Chain extended, new tip: 3bc6219cfbae66b51a3d81722e3b1135ed2e7fbb3e7ea9fa71d990310add1491 at slot 47079
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:21:59.77 UTC] Chain extended, new tip: a419a6c6aafd309903c0dcef5d83d21b8f3b31bcadb2bc2fc78fff86f2b4e1a0 at slot 47237
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:22:01.02 UTC] Chain extended, new tip: e7334bed09c631e1c57e34b94e0eed71d9f144e01deff97b34ff1a9c2a190061 at slot 47442
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:22:02.29 UTC] Chain extended, new tip: c51fd6044bfd31ab1e043350d5c5c4d9aba6b040ef99081185dd6cc4c621cbd1 at slot 47647
[relay:cardano.node.ChainDB:Notice:43] [2021-01-04 19:22:03.54 UTC] Chain extended, new tip: bfbab36bae5f0b77d3d1590c476185620ff854caee775ad8eefab84fe4c676f2 at slot 47836
```

Any idea what I could do next to find out what's going on. Is there perhaps a way to dump stack traces for active threads?

**Additional context**

CrossRef: https://github.com/tdiesler/nessus-cardano/issues/6

This seems to be a general ARM64 issue and not so much a RaspberryPi 4 one.
On x86_64 this works fine.

**Comment by AstroWa3l** [comment]
I have been trying to sync to it for 12 hours but need to keep stopping the relay and then starting it again about every hour or so to keep the sync process going. Any ideas why this would happen? My CPU is not maxed out.

**Comment by Scalextrix** [comment]
compiled version 1.25.1 from source on Pi4 8GB OS Ubuntu Server 64bit and am seeing similar issue. Compiled with cabal-install 3.4.0.0 and ghc 8.10.2, the node runs and starts to sync, after a period of about 20 minutes cardano-cli query tip --mainnet becomes unresponsive. Running cardano-node in the terminal reveals the sync appears to stop, but the node is still running full tilt, left overnight consumed 80% of RAM but no additional blocks written to disk. Restarting the cardano-node resumes sync until it fails again. cardano-cli did error on socket exhaustion one time, but cant recreate that condition reliably.

May be a localhost socket exhaustion issue?

**Comment by karknu** [comment]
You need ghc-8.10.4 for Arm in order to get the needed fixes to the haskell RTS.
With out it it is possible that cardano-node deadlocks or livelocks (as in your case).

**Comment by tdiesler** [comment]
This has been fixed in [nessus-cardano-1.25.1-rev3](https://github.com/tdiesler/nessus-cardano/releases/tag/1.25.1-rev3)
I think we can close this now.
